flytekit/clients/auth/authenticator.py

```diff
@@ -0,0 +1,228 @@
+    """
+
+    def __init__(self, endpoint: str, cfg_store: ClientConfigStore, header_key: str = None):
+        """
+        Initialize with default creds from KeyStore using the endpoint name
+        """
+        super().__init__(endpoint, header_key, KeyringStore.retrieve(endpoint))
+        self._cfg_store = cfg_store
+        self._auth_client = None
+
+    def _initialize_auth_client(self):
+        if not self._auth_client:
+            cfg = self._cfg_store.get_client_config()
+            self._set_header_key(cfg.header_key)
+            self._auth_client = AuthorizationClient(
+                endpoint=self._endpoint,
+                redirect_uri=cfg.redirect_uri,
+                client_id=cfg.client_id,
+                scopes=cfg.scopes,
+                auth_endpoint=cfg.authorization_endpoint,
+                token_endpoint=cfg.token_endpoint,
+            )
+
+    def refresh_credentials(self):
+        """ """
+        self._initialize_auth_client()
+        if self._creds:
+            """We have an access token so lets try to refresh it"""
+            try:
+                self._creds = self._auth_client.refresh_access_token(self._creds)
+                if self._creds:
+                    KeyringStore.store(self._creds)
+                return
+            except AccessTokenNotFoundError:
+                logging.warning("Failed to refresh token. Kicking off a full authorization flow.")
+                KeyringStore.delete(self._endpoint)
+
+        self._creds = self._auth_client.get_creds_from_remote()
+        KeyringStore.store(self._creds)
+
+
+class CommandAuthenticator(Authenticator):
+    """
+    This Authenticator retreives access_token using the provided command
+    """
+
+    def __init__(self, command: typing.List[str], header_key: str = None):
+        self._cmd = command
+        if not self._cmd:
+            raise ValueError("Command cannot be empty for command authenticator")
+        super().__init__(header_key)
+
+    def refresh_credentials(self):
+        """
+        This function is used when the configuration value for AUTH_MODE is set to 'external_process'.
+        It reads an id token generated by an external process started by running the 'command'.
+        """
+        if not self._cmd:
+            raise AuthenticationError("No command specified in configuration for command authentication")
+        logging.debug("Starting external process to generate id token. Command {}".format(self._cmd))
+        try:
+            output = subprocess.run(self._cmd, capture_output=True, text=True, check=True)
+        except subprocess.CalledProcessError as e:
+            logging.error("Failed to generate token from command {}".format(self._cmd))
+            raise AuthenticationError("Problems refreshing token with command: " + str(e))
+        self._creds = Credentials(output.stdout.strip())
+
+
+class ClientCredentialsAuthenticator(Authenticator):
```

Comment:
We should also add Device flow at some point...

Reply:
wdym?